examples/ecs/computed_states.rs

```diff
@@ -0,0 +1,678 @@
+struct TurboMode;
+
+impl ComputedStates for TurboMode {
+    type SourceStates = AppState;
+
+    fn compute(sources: Option<AppState>) -> Option<Self> {
+        match sources {
+            Some(AppState::InGame { turbo: true, .. }) => Some(Self),
+            _ => None,
+        }
+    }
+}
+
+// For the [`IsPaused`] state, we'll actually use an `enum` - because the difference between `Paused` and `NotPaused`
+// involve activating different systems.
+//
+// To clarify the difference, `InGame` and `TurboMode` both activate systems if they exist, and there is
+// no variation within them. So we defined them as Zero-Sized Structs.
+//
+// In contrast, pausing actually involve 3 distinct potential situations:
+// - it doesn't exist - this is when being paused is meaningless, like in the menu.
+// - it is `NotPaused` - in which elements like the movement system are active.
+// - it is `Paused` - in which those game systems are inactive, and a pause screen is shown.
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
+enum IsPaused {
+    NotPaused,
+    Paused,
+}
+
+impl ComputedStates for IsPaused {
+    type SourceStates = AppState;
+
+    fn compute(sources: Option<AppState>) -> Option<Self> {
+        // Here we convert from our [`AppState`] to all potential [`IsPaused`] versions.
+        match sources {
+            Some(AppState::InGame { paused: true, .. }) => Some(Self::Paused),
+            Some(AppState::InGame { paused: false, .. }) => Some(Self::NotPaused),
+            // If `AppState` is not `InGame`, pausing is meaningless, and so we set it to `None`.
+            _ => None,
+        }
+    }
+}
+
+// Lastly, we have our tutorial, which actually has a more complex derivation.
+//
+// Like `IsPaused`, the tutorial has a few fully distinct possible states, so we want to represent them
+// as an Enum. However - in this case they are all dependant on multiple states: the root [`TutorialState`],
+// and both [`InGame`] and [`IsPaused`] - which are in turn derived from [`AppState`].
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
+enum Tutorial {
+    MovementInstructions,
+    PauseInstructions,
+}
+
+impl ComputedStates for Tutorial {
+    // We can also use tuples of types that implement [`States`] as our [`SourceStates`].
+    // That includes other [`ComputedStates`] - though circular dependencies are not supported
+    // and will produce a compile error.
+    //
+    // We could define this as relying on [`TutorialState`] and [`AppState`] instead, but
+    // then we would need to duplicate the derivation logic for [`InGame`] and [`IsPaused`].
+    // In this example that is not a significant undertaking, but as a rule it is likely more
+    // effective to rely on the already derived states to avoid the logic drifting apart.
+    type SourceStates = (TutorialState, InGame, IsPaused);
+
+    fn compute(
+        // This complex type translates to (Option<TutorialState>, Option<InGame>, Option<IsPaused>)
```

Comment:
nit: I think the comment 'This complex type translates to (Option<TutorialState>, Option<InGame>, Option<IsPaused>)' is now outdated.